app/src/main/kotlin/com/flxrs/dankchat/DankChatViewModel.kt

```diff
@@ -1,31 +1,67 @@
 package com.flxrs.dankchat
 
 import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
 import com.flxrs.dankchat.data.ChatRepository
 import com.flxrs.dankchat.data.DataRepository
+import com.flxrs.dankchat.data.api.ApiManager
+import com.flxrs.dankchat.preferences.DankChatPreferenceStore
+import com.flxrs.dankchat.utils.extensions.withoutOAuthSuffix
 import dagger.hilt.android.lifecycle.HiltViewModel
+import kotlinx.coroutines.channels.BufferOverflow
+import kotlinx.coroutines.flow.*
+import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 @HiltViewModel
 class DankChatViewModel @Inject constructor(
     private val chatRepository: ChatRepository,
+    private val dankChatPreferenceStore: DankChatPreferenceStore,
+    private val apiManager: ApiManager,
     dataRepository: DataRepository,
 ) : ViewModel() {
 
     val commands = dataRepository.commands
     var started = false
         private set
 
+    private val _currentUserName = MutableStateFlow<String?>(null)
+    val currentUserName: StateFlow<String?> get() = _currentUserName.asStateFlow()
+
+    private val _validationError = MutableSharedFlow<OAuthValidationError>(extraBufferCapacity = 1, onBufferOverflow = BufferOverflow.DROP_OLDEST)
+    val validationError get() = _validationError.asSharedFlow()
+
     fun init(name: String, oAuth: String, channels: List<String>, tryReconnect: Boolean) {
         if (tryReconnect && started) {
             chatRepository.reconnectIfNecessary()
         } else {
             started = true
-            chatRepository.connectAndJoin(name, oAuth, channels)
+            viewModelScope.launch {
+                val token = oAuth.withoutOAuthSuffix
+
+                var nameToUpdate = name // fallback to old name
+                runCatching {
```

Comment:
This can be refactored a bit to not have the `var` by using `getOrElse` and moving some of the success logic inside the `runCatching`:
```kotlin
val token = oAuth.withoutOAuthSuffix

val validatedName = runCatching {
    val validationResult = apiManager.validateUser(token)
    if (validationResult== null && token.isNotEmpty()) {
        _validationError.emit(OAuthValidationError.OAuthTokenInvalid)
    }

    validationResult?.login
}.getOrElse {
    // Connection failure
    _validationError.emit(OAuthValidationError.OAuthValidationFailure)
    null
}
val nameToUpdate = validatedName ?: name // fallback to old name
_currentUserName.value = nameToUpdate
dankChatPreferenceStore.userName = nameToUpdate
chatRepository.connectAndJoin(nameToUpdate, oAuth, channels)
```